plutus-core/plutus-ir/src/PlutusIR/Transform/CommuteFnWithConst.hs

```diff
@@ -0,0 +1,130 @@
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE LambdaCase       #-}
+{-# LANGUAGE TypeApplications #-}
+
+module PlutusIR.Transform.CommuteFnWithConst
+  (commuteFnWithConst
+  , commuteDefaultFun)
+  where
+
+import Control.Lens (over)
+import Data.Typeable (Typeable, eqT)
+import PlutusCore.Default
+import PlutusIR.Core.Plated (termSubterms)
+import PlutusIR.Core.Type (Term (Apply, Builtin, Constant))
+
+{- | Commute such that constants are the first arguments. Consider:
+
+(1)    equalsInteger 1 x
+
+(2)    equalsInteger x 1
+
+We have unary application, so these are two partial applications:
+
+(1)    (equalsInteger 1) x
+
+(2)    (equalsInteger x) 1
+
+With (1), we can share the `equalsInteger 1` node, and it will be the same across any place where
+we do this.
+
+With (2), both the nodes here include x, which is a variable that will likely be different in other
+invocations of `equalsInteger`. So the second one is harder to share, which is worse for CSE.
+
+So commuting `equalsInteger` so that it has the constant first both a) makes various occurrences of
+`equalsInteger` more likely to look similar, and b) gives us a maximally-shareable node for CSE.
+
+This applies to any commutative builtin function that takes constants as arguments, although we
+might expect that `equalsInteger` is the one that will benefit the most.
+Plutonomy only commutes `EqualsInteger` in their `commEquals`.
+-}
+
+isConstant :: Term tyname name uni fun a -> Bool
+isConstant Constant{} = True
+isConstant _          = False
+
+commuteDefaultFun ::
+    forall tyname name uni a.
+    Term tyname name uni DefaultFun a ->
+    Term tyname name uni DefaultFun a
+commuteDefaultFun = over termSubterms commuteDefaultFun . localCommute
+  where
+    localCommute tm@(Apply ann (Apply ann1 (Builtin annB fun) x) y) =
+      case (isCommutativeWithConstant fun, isConstant x, isConstant y) of
```

Comment:
Use guards!

Reply:
And for the constraint on `y` you can just inline the pattern into the one on the line above.